pkg/cmdline/cmd.go

```diff
@@ -0,0 +1,124 @@
+)
+
+// CommandManager holds root command or first parent
+// and can stores group of command
+type CommandManager struct {
+	rootCmd   *cobra.Command
+	groupCmds map[string][]*cobra.Command
+	errPool   []error
+	fm        *flagManager
+}
+
+// NewCommandManager instantiates a CommandManager
+func NewCommandManager(rootCmd *cobra.Command) *CommandManager {
+	if rootCmd == nil {
+		panic("nil root command passed")
+	}
+	cm := &CommandManager{
+		rootCmd:   rootCmd,
+		groupCmds: make(map[string][]*cobra.Command),
+		errPool:   make([]error, 0),
+		fm:        newFlagManager(),
+	}
+	return cm
+}
+
+func (m *CommandManager) pushError(f string, a ...interface{}) {
+	m.errPool = append(m.errPool, fmt.Errorf(f, a...))
+}
+
+// GetError returns the error pool
+func (m *CommandManager) GetError() []error {
+	return m.errPool
+}
+
+// RegisterCmd registers a child command for the root command
+func (m *CommandManager) RegisterCmd(cmd *cobra.Command, interspersed bool) {
+	// panic here because it's a misuse of API and generally from
+	// global context or init() functions
+	if cmd == nil {
+		panic("nil command passed")
+	}
+	m.rootCmd.AddCommand(cmd)
+	cmd.Flags().SetInterspersed(interspersed)
+}
+
+// RegisterSubCmd registers a child command for parent command given as argument
+func (m *CommandManager) RegisterSubCmd(parentCmd, childCmd *cobra.Command, interspersed bool) {
+	// panic here because it's a misuse of API and generally from
+	// global context or init() functions
+	if parentCmd == nil {
+		panic("nil parent command passed")
+	} else if childCmd == nil {
+		panic("nil child command passed")
+	}
+	parentCmd.AddCommand(childCmd)
+	childCmd.Flags().SetInterspersed(interspersed)
+}
+
+// SetCmdGroup creates a group of commands identified by name
+func (m *CommandManager) SetCmdGroup(name string, cmds ...*cobra.Command) {
+	m.groupCmds[name] = make([]*cobra.Command, 0)
+	for _, c := range cmds {
+		if c == nil {
+			panic("nil command passed")
+		}
+		m.groupCmds[name] = append(m.groupCmds[name], c)
+	}
+}
+
+// GetCmdGroup returns group of commands corresponding to name
+func (m *CommandManager) GetCmdGroup(name string) []*cobra.Command {
+	return m.groupCmds[name]
+}
+
+// GetRootCmd returns the root command
+func (m *CommandManager) GetRootCmd() *cobra.Command {
+	return m.rootCmd
+}
+
+// GetCmd returns the named command associated with root command
+func (m *CommandManager) GetCmd(name string) *cobra.Command {
+	for _, c := range m.rootCmd.Commands() {
+		if c.Name() == name {
+			return c
+		}
+	}
+	return nil
+}
+
+// GetSubCmd return the named command associated with parent command given as argument
+func (m *CommandManager) GetSubCmd(parentCmd *cobra.Command, name string) *cobra.Command {
+	for _, c := range parentCmd.Commands() {
+		if c.Name() == name {
+			return c
+		}
+	}
+	return nil
+}
+
+// RegisterCmdFlag registers a flag for a command
+func (m *CommandManager) RegisterCmdFlag(flag *Flag, cmds ...*cobra.Command) {
```

Comment:
I was wondering why you had cmd as the second argument...

Since the common case is to register a flag for a single command, wouldn't it be more natural to have `RegisterCmdFlag(cmd *cobra.Command, flag *Flag)`?

Reply:
Because cmds is a variadic argument

Reply:
Look at `cmd/internal/cli/action_flags.go` and you will see it's useful. I can rename it to `RegisterFlagForCmd(flag, cmds...)` to be more natural, but I really want to keep variadic command argument